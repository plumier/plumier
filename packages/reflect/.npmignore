.vscode
coverage
scripts
src 
test
.gitignore
.npmignore
.travis.yml
package-lock.json
tsconfig.build.json
tsconfig.json
yarn-error.log
yarn.lock