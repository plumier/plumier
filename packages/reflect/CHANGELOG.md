# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.1.1](https://github.com/plumier/plumier/compare/v1.1.0...v1.1.1) (2023-01-03)


### Bug Fixes

* **reflect:** Fix issue when reflect async function inside object ([#1033](https://github.com/plumier/plumier/issues/1033)) ([e2be8ca](https://github.com/plumier/plumier/commit/e2be8ca167e9fc3f959d5f6b26d0aa6d8cec49fb))
* **reflect:** Fix issue when reflect function inside object ([#1032](https://github.com/plumier/plumier/issues/1032)) ([719d23a](https://github.com/plumier/plumier/commit/719d23ab41785c02f0e851a2074c86ee0afe6dd1))





# [1.1.0](https://github.com/plumier/plumier/compare/v1.0.6...v1.1.0) (2022-12-19)


### Bug Fixes

* **reflect:** Fix reflect unable to inspect lambda function parameters ([#1030](https://github.com/plumier/plumier/issues/1030)) ([2cc9e82](https://github.com/plumier/plumier/commit/2cc9e82601b70703218da8e09bd38543fb09332e))
* **reflect:** Remove unnecessary base class when create class using property definition ([#1031](https://github.com/plumier/plumier/issues/1031)) ([9134b6d](https://github.com/plumier/plumier/commit/9134b6d787eafaff4ba5a17629b28d5178543055))


### Features

* Add feature to reflect function and lambda function ([#1029](https://github.com/plumier/plumier/issues/1029)) ([f65ec5b](https://github.com/plumier/plumier/commit/f65ec5baa14af9b34793a2ab74f440233f84c6d7))
* **reflect:** Added support for static methods and properties reflection ([#1023](https://github.com/plumier/plumier/issues/1023)) ([f8ee76d](https://github.com/plumier/plumier/commit/f8ee76dbe3a8543f556dfc3d3ac2468376cff482))





## [1.0.6](https://github.com/plumier/plumier/compare/v1.0.5...v1.0.6) (2021-10-04)

**Note:** Version bump only for package @plumier/reflect





## [1.0.5](https://github.com/plumier/plumier/compare/v1.0.4...v1.0.5) (2021-07-25)

**Note:** Version bump only for package @plumier/reflect





## [1.0.3](https://github.com/plumier/plumier/compare/v1.0.2...v1.0.3) (2021-06-28)


### Bug Fixes

* Cache class parser to speed up class introspection ([#970](https://github.com/plumier/plumier/issues/970)) ([aca5751](https://github.com/plumier/plumier/commit/aca5751382bb81bf264df5fbc7dc27df15943874))





## [1.0.2](https://github.com/plumier/plumier/compare/v1.0.0...v1.0.2) (2021-06-06)

**Note:** Version bump only for package @plumier/reflect





## 1.0.1 (2021-05-18)

**Note:** Version bump only for package @plumier/reflect
