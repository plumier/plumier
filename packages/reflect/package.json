{
  "name": "@plumier/reflect",
  "version": "1.0.0-rc.11",
  "description": "Plumier reflection module",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "scripts": {
    "test": "jest",
    "compile": "tsc -p tsconfig.build.json",
    "clean": "rm -rf src/*.js && rm -rf test/*.js && rm -rf test/**/*.js && rm -rf lib"
  },
  "keywords": [
    "Reflection",
    "Reflect",
    "Metadata"
  ],
  "author": "Ketut Sandiarsa",
  "license": "MIT",
  "dependencies": {
    "@types/acorn": "4.0.5",
    "acorn": "8.1.1",
    "reflect-metadata": "^0.1.13"
  },
  "devDependencies": {
    "@types/jest": "^26.0.22",
    "@types/node": "^14.14.41",
    "coveralls": "^3.1.0",
    "jest": "^26.6.3",
    "ts-jest": "^26.5.4",
    "tslib": "^2.2.0",
    "typescript": "^4.2.4"
  },
  "bugs": {
    "url": "https://github.com/plumier/plumier/issues"
  },
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "url": "https://github.com/plumier/plumier"
  }
}
