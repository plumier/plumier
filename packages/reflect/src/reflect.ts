import { ignore, noop, parameterProperties, type } from "./decorators"
import { createClass, metadata, useCache } from "./helpers"
import { parseFunction } from "./parser"
import { Class, ClassReflection, ObjectReflection, Reflection } from "./types"
import { memberVisitors as v, walkTypeMembersRecursive, walkReflectionMembers } from "./walker"

interface TraverseContext {
    path: any[]
}

type Visitor<Ref, Ctx, Ret> = (r: Ref, c: Ctx) => Ret

function pipe<Ref, Ctx, Ret>(visitors: Visitor<Ref, Ctx, Ret>[]): Visitor<Ref, Ctx, Ret> {
    return (value, ctx) => visitors.reduce((a, b) => !!a ? b(a, ctx) : a, value as any)
}

function reflectClass(target: Class): ClassReflection {
    const ref = walkTypeMembersRecursive(target, {
        target, memberVisitor: pipe([
            // add typescript design types information
            v.addsDesignTypes,
            // add metadata decorators
            v.addsDecorators,
            // add datatype info specified by @type decorator
            v.addsTypeOverridden,
            // add parameter properties
            v.addsParameterProperties,
            // add generic type information
            v.addsGenericOverridden,
            // add typeClassification information
            v.addsTypeClassification,
            // remove @ignore decorator
            v.removeIgnored,
        ]),
        classPath: []
    })
    return walkReflectionMembers(ref, {
        classPath: [], target, parent:ref, memberVisitor: pipe([
            v.addsApplyToDecorator,
            // remove generic type overridden, because it may cause issue for addsTypeOverridden
            v.removeGenericTypeOverridden,
            // add datatype info specified by @type decorator
            v.addsTypeOverridden,
            // add typeClassification information
            v.addsTypeClassification,
        ])})
}

function traverseObject(fn: any, name: string, ctx: TraverseContext): Reflection | undefined {
    if (fn === undefined || fn === null) return
    if (Array.isArray(fn)) return
    if (typeof fn === "object") {
        // CIRCULAR: return immediately
        if (ctx.path.some(x => x === fn)) return
        return reflectObject(fn, name, { path: ctx.path.concat(fn) })
    }
    if (typeof fn === "function" && metadata.isConstructor(fn))
        return reflectClass(fn)
    if (typeof fn === "function")
        return parseFunction(fn)
}

function reflectObject(object: any, name: string, ctx: TraverseContext): ObjectReflection {
    return {
        kind: "Object", name,
        members: Object.keys(object).map(x => traverseObject(object[x], x, ctx)).filter((x): x is Reflection => !!x)
    }
}

function reflectModuleOrClass(opt: string | Class) {
    if (typeof opt === "string") {
        return reflectObject(require(opt), "module", { path: [] })
    }
    else {
        return reflectClass(opt)
    }
}


// --------------------------------------------------------------------- //
// ------------------------------- CACHE ------------------------------- //
// --------------------------------------------------------------------- //

interface ReflectOption {
    /**
     * Flush cached current module/class metadata before processing. 
     */
    flushCache?: true
}

const cacheStore = new Map<string | Class, ClassReflection | ObjectReflection>()
const reflectCached = useCache(cacheStore, reflectModuleOrClass, x => x)

/**
 * Reflect module
 * @param path module name
 */
function reflect(path: string, opt?: Partial<ReflectOption>): ObjectReflection

/**
 * Reflect class
 * @param classType Class 
 */
function reflect(classType: Class, opt?: Partial<ReflectOption>): ClassReflection

function reflect(pathOrClass: string | Class, opt?: Partial<ReflectOption>): ClassReflection | ObjectReflection {
    if (opt?.flushCache)
        cacheStore.delete(pathOrClass)
    return reflectCached(pathOrClass)
}

function flush(pathOrClass?: string | Class) {
    if (pathOrClass)
        cacheStore.delete(pathOrClass)
    else
        cacheStore.clear()
}

// --------------------------------------------------------------------- //
// ----------------------------- DECORATORS ---------------------------- //
// --------------------------------------------------------------------- //

/**
 * Decorator that do nothing, intended to be able to inspect data type
 */
reflect.noop = noop

/**
 * Ignore member from metadata generated
 */
reflect.ignore = ignore

/**
 * Override type definition information. Useful to add type definition for some data type that is erased 
 * after transfile such as Partial<Type> or ReadOnly<Type>
 * 
 * If applied to parameter it will override the parameter type
 * 
 * If applied to property it will override the property type
 * 
 * if applied to method it will overrid the method return value
 * @param type The type overridden
 * @param info Additional information about type (readonly, partial etc)
 */
reflect.type = type

/**
 * Create class dynamically
 */
reflect.create = createClass

/**
 * Mark all constructor parameters as properties
 */
reflect.parameterProperties = parameterProperties

/**
 * Flush metadata cache, call without parameter to clear all
 */
reflect.flush = flush

export { reflect }
