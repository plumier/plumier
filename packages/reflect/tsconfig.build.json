{
    "extends": "./tsconfig.json",
    "compilerOptions": {
        "target": "es6",
        "outDir": "lib",
        "declaration": true,
        "inlineSourceMap": false
    },
    "exclude": [
        "test",
        "lib"
    ]
}