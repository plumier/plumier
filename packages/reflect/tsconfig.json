{
  "compilerOptions": {
    "target": "es2017",
    "module": "commonjs",
    "strict": true,
    "inlineSourceMap": true,
    "esModuleInterop": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true       
  },
  "exclude": [
    "node_modules"
  ]
}