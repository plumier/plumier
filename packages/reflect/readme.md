# Reflect: A JavaScript/TypeScript Reflection Library

A reflection library that extracts JavaScript/TypeScript object into an object graph containing object preferences and metadata information


## Companies Using Reflect

[![HP Inc](https://avatars.githubusercontent.com/u/9728779?s=200)](https://github.com/HPInc)

## Example Usage

```typescript
import reflect from "@plumier/reflect"

class Awesome{
    awesome(){}
}

class MyAwesomeClass extends Awesome {
    @decorateMethod({ cache: "20s" })
    myAwesomeMethod(stringPar:string, numberPar:number): number {
        return Math.random()
    }
}

const obj = reflect(MyAwesomeClass)
```

Result of `obj` variable above is like below
    
```javascript
{
    kind: 'Class',
    name: 'MyAwesomeClass',
    type: MyAwesomeClass,
    decorators: [],
    properties: [],
    ctor: {
        kind: 'Constructor',
        name: 'constructor',
        parameters: []
    },
    methods: [
        {
            kind: 'Method',
            name: 'myAwesomeMethod',
            parameters: [
                {
                    kind: 'Parameter',
                    name: 'stringPar',
                    type: String,
                    decorators: [],
                    properties: {}
                },
                {
                    kind: 'Parameter',
                    name: 'numberPar',
                    type: Number,
                    decorators: [],
                    properties: {}
                }
            ],
            decorators: [{ cache: '20s' }],
            returnType: Number,
        },
        {
            kind: 'Method',
            name: 'awesome',
            parameters: [],
            decorators: [],
            returnType: undefined,
        }
    ]
}
```

## Features

- [x] Inspect function
- [x] Inspect module or file
- [x] Inspect class
- [x] Inspect class with inheritance
- [x] Inspect getter and setter
- [x] Inspect methods
- [x] Inspect parameters
- [x] Supported inspect destructured parameter
- [x] Supported inspect rest parameter
- [x] Supported inspect parameter with complex default value
- [x] (TypeScript only) Inspect metadata using decorators
- [x] (TypeScript only) Inspect parameter properties
- [x] (TypeScript only) Inspect type information 
- [x] (TypeScript only) Configurable decorator (inheritable / allow multiple)
- [x] (TypeScript only) Generic class inheritance

## TypeScript Requirement
To be able to inspect type information its required to enable configuration below in `tsconfig.json`

```json
{
  "compilerOptions": {
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true           
  }
}
```

## Inspect Type Information
TypeScript when `emitDecoratorMetadata` enabled, TypeScript will add type information during compile time. This make it possible to extract typescript type information during runtime. 

> **CAVEAT**
>  
> TypeScript `emitDecoratorMetadata` has some limitation. 
> * Declaration should have at least one decorator to get metadata (type information)
> * Array element type information not included. 
> * Generic type information not included.

Based on limitation above its required at least have one decorator to make reflection to be able to extract type information: 

```typescript 
import reflect from "@plumier/reflect"

class Awesome {
    @reflect.noop()
    awesome(multiply:number): number { return 1 }
}

const obj = reflect(Awesome)
```

reflection will be able to get type information of the method's return type and parameters of the `awesome` method above. Note that we applied `@reflect.noop()` decorator on the `awesome` method. `@reflect.noop()` does nothing except to force TypeScript to emit metadata information.

```typescript 
import reflect from "@plumier/reflect"

class Awesome {
    @reflect.noop()
    aweProperty:number
}

const obj = reflect(Awesome)
```

Above code showing that we able to get type information of a property. 

```typescript 
import reflect from "@plumier/reflect"

@reflect.noop()
class Awesome {
    constructor(multiply:number){}
}

const obj = reflect(Awesome)
```
Above code showing that we able to get type information of parameters of the constructor, by applying decorator on the class level. 

## Inspect Array and Generic Type
To get type information of an Array and Generic type its required to use `@reflect.type()` decorator and provide the type on the callback parameter.

```typescript 
import reflect from "@plumier/reflect"

class Awesome {
    @reflect.type([Number])
    awesome(multiply:number): Array<number> {}
}

const obj = reflect(Awesome)
```

Above code showing that we able to get method's return type information by providing `@reflect.type([Number])`. Note that the `[Number]` is an array of `Number`. 

```typescript 
import reflect from "@plumier/reflect"

class Option { 
    data:string
}

class Awesome {
    @reflect.type(Option)
    awesome(multiply:number): Partial<Option> {}
}

const obj = reflect(Awesome)
```

We will be able to get generic type information such as `Partial`, `Required` etc by applying `@reflect.type()` like above. 

> If you receive `ReferenceError: <type name> is not defined` error, thats mean you use `@reflect.type(MyType)` prior than its declaration. To solve this issue you can use callback `@reflect.type(x => MyType)` or `@reflect.type(x => [MyType])` for array type.

## Inspect Generic Class Information
With above trick its possible to get generic type information in a generic class members with some extra configuration below

```typescript
@generic.parameter("T", "U")
class SuperAwesome<T, U> {
    awesome(@reflect.type("T") par: T, @reflect.type("U") par2: U) {}
}

@generic.argument(Number, String)
class Awesome extends SuperAwesome<Number, String> { }

const obj = reflect(Awesome)
```

Above code showing that we add a specialized decorator `@generic.parameter()` to define generic type parameter. We also defined data type of the generic parameters using `@reflect.type()` decorator. Next on the inherited class we specify `@generic.argument()` to define types replace the generic template. Note that the order of the parameter on `@generic.parameter()` and `@generic.argument()` is important.

## Inspect Parameter Properties
TypeScript has parameter properties feature, which make it possible to use constructor parameter as property. reflection able to extract parameter properties type information by using `@reflect.parameterProperties()` decorator.

```typescript 
import reflect from "@plumier/reflect"

@reflect.parameterProperties()
class Awesome {
    constructor(public multiply:number){}
}

const obj = reflect(Awesome)
```

## Metadata Decorator 
Reflection able to extract classes/methods/properties/parameters decorated with predefined metadata decorators. There are predefined decorators should be use to be able for reflection to inspect the decorators

| Decorator           | Description                                                                          |
| ------------------- | ------------------------------------------------------------------------------------ |
| `decorateClass`     | Decorate class with object specified in parameter                                    |
| `decorateProperty`  | Decorate property with object specified in parameter                                 |
| `decorateMethod`    | Decorate method with object specified in parameter                                   |
| `decorateParameter` | Decorate parameter with object specified in parameter                                |
| `decorate`          | Decorate any (class, property, method, parameter) with object specified in parameter |
| `mergeDecorator`    | Merge multiple decorators into one, useful on creating custom decorator               |

Example usage

```typescript
import { decorateClass, decorateProperty } from "@plumier/reflect"

@decorateClass({ message: "hello world" })
class Awesome {
    @decorateProperty({ message: "awesome!" })
    awesome: number = 10
}
```

Parameter passed on each decorator can be any object contains value, methods etc, those value will be returned when the class metadata extracted.

Create your own custom decorator by creating function returns decorator above

```typescript
import { decorateMethod } from "@plumier/reflect"

// create custom decorator
function cache(duration:number){
    return decorateMethod({ type: "Cache", duration })
}

class Awesome{
    @cache() // use it like usual decorator
    awesome(){}
}
```

Use `mergeDecorator` to combine multiple decorator on custom decorator


```typescript
import { decorateMethod, mergeDecorator } from "@plumier/reflect"

// create custom decorator
function cacheAndDelay(duration:number){
    return mergeDecorator([
        decorateMethod({ type: "Cache", duration }), 
        decorateMethod({ type: "Delay", duration })
    ])
}
```

## Decorator Option (Inheritance, Allow Multiple)
Decorator can be further configured to match the behavior you need like below.

```typescript
decorateMethod(<data>, <option>)
```

Option is a simple object with properties: 

* `inherit` `Boolean` If `false` decorator will not be merged on the derived class. Default `true` 
* `allowMultiple` `Boolean` If `false` throw error when multiple decorator applied on class. Also when set `false` will prevent super class decorator being merged into derived class when already exists. When set `false`, decorator required to have `DecoratorId` property to identify the similar decorator.
* `applyTo` `String` or `String[]` apply decorator into the specified properties or methods.
* `removeApplied` `Boolean` Remove applied decorator using `applyTo` on the class scope. Default `true`.


Example disable decorator inheritance

```typescript
@decorateClass({ log:true }, { inherit: false })
class Awesome {
    awesome(){}
}

// { log: true} will not inherited on this class
class IamAwesome extends Awesome{ }
```

Example disable multiple decorator on inheritance

```typescript
import { DecoratorId, decorateClass } from "@plumier/reflect"

function log(){
    return @decorateClass({ [DecoratorId]: "logging",  log:true }, { allowMultiple: false })
}

@log()
class Awesome {
    awesome(){}
}

// parent decorator will not merged
// guaranteed derived class only have single decorator with specific ID
@log()
class IamAwesome extends Awesome{ }
```

## Apply Decorator from Class
Sometime its required to add decorator into specific methods or properties from class. This is useful when you want to decorate specific super class method from derived class. Use `applyTo` and `removeApplied` properties on decorator option to control decorator to be copied into the properties or methods.


```typescript
import { decorateClass } from "@plumier/reflect"

@decorateClass({ message: "hello world" }, { applyTo: "myMethod" })
class Awesome {
    myMethod(){}
}
```

By specifying `{ applyTo: "myMethod" }` on the decorator option will make the decorator copied into the `myMethod` method, while the decorator removed from the class. 

```typescript
import { decorateClass } from "@plumier/reflect"

@decorateClass({ message: "hello world" }, { applyTo: "myMethod", removeApplied: false })
class Awesome {
    myMethod(){}
}
```

Above code will copy decorator into `myMethod` method but keep the decorator on the class.

## Flush Cache 
By default reflect process cached globally for performance reason. But in some case if you modify the class preferences by adding a new decorator etc, your new update will not returned until you flush the cache.

```typescript
reflect.flush(Type)
```