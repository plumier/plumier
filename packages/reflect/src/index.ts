import "reflect-metadata"

export * from "./types"
export * from "./decorators"
export * from "./helpers"
import { reflect } from "./reflect"
export { reflect }
export default reflect
