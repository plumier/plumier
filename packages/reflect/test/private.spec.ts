import reflect from "../src";


describe("Private Decorator", () => {
    it("Should be able to ignore get set", () => {
        class DummyClass {
            @reflect.ignore()
            get data() { return 1 }
        }
        
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should be able to ignore method", () => {
        class DummyClass {
            @reflect.ignore()
            data() { return 1 }
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })

    it("Should be able to ignore parameter properties", () => {
        @reflect.parameterProperties()
        class DummyClass {
            constructor(public data:string, @reflect.ignore() public id:string){}
        }
        const meta = reflect(DummyClass)
        expect(meta).toMatchSnapshot()
    })
})