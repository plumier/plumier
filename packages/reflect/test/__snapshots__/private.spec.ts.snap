// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Private Decorator Should be able to ignore get set 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [],
  "name": "DummyClass",
  "properties": Array [],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Private Decorator Should be able to ignore method 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [],
  "name": "DummyClass",
  "properties": Array [],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;

exports[`Private Decorator Should be able to ignore parameter properties 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [
      Object {
        "decorators": Array [],
        "fields": Object {},
        "index": 0,
        "kind": "Parameter",
        "name": "data",
        "type": String,
        "typeClassification": "Primitive",
      },
      Object {
        "decorators": Array [
          Object {
            "kind": "Ignore",
            Symbol(tinspector:decoratorId): Symbol(ignore),
            Symbol(tinspector:decoratorOption): Object {
              "allowMultiple": false,
              "applyTo": Array [],
              "inherit": true,
              "removeApplied": true,
            },
          },
        ],
        "fields": Object {},
        "index": 1,
        "kind": "Parameter",
        "name": "id",
        "type": String,
        "typeClassification": "Primitive",
      },
    ],
  },
  "decorators": Array [
    Object {
      "type": "ParameterProperties",
      Symbol(tinspector:decoratorId): Symbol(paramProp),
      Symbol(tinspector:decoratorOption): Object {
        "allowMultiple": false,
        "applyTo": Array [],
        "inherit": true,
        "removeApplied": true,
      },
    },
  ],
  "kind": "Class",
  "methods": Array [],
  "name": "DummyClass",
  "properties": Array [
    Object {
      "decorators": Array [],
      "get": undefined,
      "index": 0,
      "isParameter": true,
      "kind": "Property",
      "name": "data",
      "set": undefined,
      "type": String,
      "typeClassification": "Primitive",
    },
  ],
  "super": Object,
  "type": DummyClass,
  "typeClassification": "Class",
}
`;
